FIX: Imputer fix for indicator_matrix in add_indicator_features option

maniteja123 opened the pull request:
#### Reference Issue

Fixes #6767 
#### What does this implement/fix? Explain your changes.

Imputer `add_indicator_features` depends on the `fit` input for the indicator matrix and should be irrespective of the transform input.
#### Any other comments?

I have done many changes in the code, but the tests are passing. So please let me know if I changed something incorrectly.

maniteja123 committed:
FIX: Modify the indicator features functionality

jnothman commented on doc/modules/preprocessing.rst:
```diff
@@ -474,11 +474,11 @@ then the shape of the transformed input is
     >>> print(imp_with_in.transform(X))                           # doctest: +ELLIPSIS
-    [[ 4.          2.          1.          0.        ]
-     [ 6.          3.66666667  0.          1.        ]
+    [[ 4.          2.          0.          0.        ]
```
What??? No, the indicator features need to correspond to the `transform` input. The `X.shape[0]` does not need to be the same for `fit` and `transform`.

jnothman replied:
Clearly, you were right to ask questions because you've still missed the point. Have you *used* scikit-learn's transformers? Do you understand their API design? Do you understand what this `add_indicator_features` is for?

Yes, we made an error in reviewing this PR, but not one as glaring as to get this wrong!

maniteja123 replied:
Apologies!! This is really wrong. I have accidentally pushed the wrong commit from my other machine. I totally understand the API design and that the the fix is just about the shape of transformed output being consistent for different inputs. I will revert everything correctly. 

But one final question please - now indicator features will have a column for each feature regardless of that feature having missing values in the current transform input. But the imputed_features_ will have only features with at least one missing value. The reason is I can't do X[:, expected_imputer_features] because imputed_features won't have that feature but the indicator features will still have it. This is the only thing I suppose I didn't ask about. 

Sorry that I actually don't know what exactly how this indicator features help but I understand that they provide information about the missing data and thus can help improve the accuracy. And in case if you feel I am still not getting it right, please do let me know if it will be better for someone else to do this.

jnothman replied:
I'd like you to have a go, but showing what you've got early - and writing tests to show your understanding - helps make explicit checkpoints.

I was thinking a couple of hours ago about the case where a feature has missing values at `transform` time but not at `fit`. I do think this is a possible case, but I also think it's silly to add an indicator column for each of thousands of features not requiring imputation. We could actually allow the user to specify which features may require imputation (and hence both statistics and indicator matrix), with a default to infer them from the training data. What do you think?

maniteja123 replied:
Thank you, will definitely do that. 

The idea of user specifying seems good but would it always be the case that the user is aware of the features with missing data in the whole training data ?. And even if they do, will it be consistent for all the samples ?

Also if we proceed to add indicator features for every feature irrespective of imputation, something like variance threshold can remove such indicators features that correspond to features with no missing values right ? Or maybe give an option specifically for the indicator features to have all features or just imputed features ?

jnothman replied:
Under shuffled cross-validation, you can't always be sure what's in your training and what's in your test. That's why allowing them to specify to Imputer may be important. However, to take back what I said, the indicator features won't be any use if they're not seen in training, so it's probably valid not to add a parameter to specify which features.

maniteja123 replied:
Yeah I get the point of not having it since it should be the case where you know what features are missing in training and testing data. So it is better to proceed with adding indicator features for all the features?

maniteja123 commented:
Sorry the comments went to the outdated diff. I hope to have at least done it somewhat in the correct direction. The tests will fail right now because ``imputed_features_`` has only imputed features while the indicator features has one for all the features irrespective of whether they are imputed or not. 

Also there is a possibility of your case when input to ``fit`` doesn't have missing value for a feature but ``transform`` input might have. How should I proceed with this issue ? Thanks.

jnothman commented:
No, I think it's better to proceed as we were: indicator features only for
those features with something (but not everything) missing at training
time. Without those properties, the imputer is not much value at transform
time.

On 15 May 2016 at 19:27, Maniteja Nandana <notifications@github.com> wrote:

> Sorry the comments went to the outdated diff. I hope to have at least done
> it somewhat in the correct direction. The tests will fail right now because
> imputed_features_ has only imputed features while the indicator features
> has one for all the features irrespective of whether they are imputed or
> not.
>
> Also there is a possibility of your case when input to fit doesn't have
> missing value for a feature but transform input might have. How should I
> proceed with this issue ? Thanks.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6782#issuecomment-219275089>
>

jnothman commented:
If there is no missing value at fit time, but there is at transform, do we
currently impute it? I think at this stage, we should not return an
indicator column for that feature, on the basis that there is no training
data to learn its from...

On 15 May 2016 at 21:12, Joel Nothman <joel.nothman@gmail.com> wrote:

> No, I think it's better to proceed as we were: indicator features only for
> those features with something (but not everything) missing at training
> time. Without those properties, the imputer is not much value at transform
> time.
>
> On 15 May 2016 at 19:27, Maniteja Nandana <notifications@github.com>
> wrote:
>
>> Sorry the comments went to the outdated diff. I hope to have at least
>> done it somewhat in the correct direction. The tests will fail right now
>> because imputed_features_ has only imputed features while the indicator
>> features has one for all the features irrespective of whether they are
>> imputed or not.
>>
>> Also there is a possibility of your case when input to fit doesn't have
>> missing value for a feature but transform input might have. How should I
>> proceed with this issue ? Thanks.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/6782#issuecomment-219275089>
>>
>
>

jnothman commented:
*to learn its meaning from

On 15 May 2016 at 21:13, Joel Nothman <joel.nothman@gmail.com> wrote:

> If there is no missing value at fit time, but there is at transform, do we
> currently impute it? I think at this stage, we should not return an
> indicator column for that feature, on the basis that there is no training
> data to learn its from...
>
> On 15 May 2016 at 21:12, Joel Nothman <joel.nothman@gmail.com> wrote:
>
>> No, I think it's better to proceed as we were: indicator features only
>> for those features with something (but not everything) missing at training
>> time. Without those properties, the imputer is not much value at transform
>> time.
>>
>> On 15 May 2016 at 19:27, Maniteja Nandana <notifications@github.com>
>> wrote:
>>
>>> Sorry the comments went to the outdated diff. I hope to have at least
>>> done it somewhat in the correct direction. The tests will fail right now
>>> because imputed_features_ has only imputed features while the indicator
>>> features has one for all the features irrespective of whether they are
>>> imputed or not.
>>>
>>> Also there is a possibility of your case when input to fit doesn't have
>>> missing value for a feature but transform input might have. How should
>>> I proceed with this issue ? Thanks.
>>>
>>> —
>>> You are receiving this because you commented.
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/scikit-learn/scikit-learn/pull/6782#issuecomment-219275089>
>>>
>>
>>
>

maniteja123 commented:
The statistics are calculated at fit time except for features with all missing values for axis=0. So if feature has missing value in transform it does get imputed. But if I am not mistaken now depending on input to transform, the shape of indicator features matrix will again vary if we go with something but not everything.

jnothman commented:
Then I'm sure you're mistaken. Give me an example; a test case?

jnothman commented:
I'm not sure what you mean by your "except for", either.

maniteja123 commented:
Sorry, I meant it will be ``nan`` for features with all missing values. And a test case, if we take ``test_indicator_features_shape``, suppose we fit on X1. When X1 is transformed, the shape is (5, 6) [5, 3+3] but transformed X2 will be (5, 5)[5, 3+2] right ? Since we add indicator features only for features with some missing values and in X2 feature 3 has no missing value. Am I understanding it wrong ?

jnothman commented:
The golden rule of **every call to transform must output the same number of features** says you must be wrong. No, we add indicator features according to missing values at `fit` time, not `transform` time.

maniteja123 commented:
Sorry won't bother you more. I will try to show an example of what I understand here.

```
>>> X1 = np.array([
...     [-1,  -1,   2,   3],
...     [4,  -1,   6,  -1],
...     [8,  -1,  10,  11],
...     [12,  -1,  -1,  15],
...     [16,  -1,  18,  19]
... ])
>>> 
>>> X2 = np.array([
...     [-1,  -1,   1,   3],
...     [4,  -1,   0,  -1],
...     [8,  -1,   1,  0],
...     [0,  -1,   0,  15],
...     [16,  -1,   1,  19]
... ])

>>> imputer = Imputer(missing_values=-1, strategy='mean', axis=0,add_indicator_features=True)
>>> imputer.fit(X1)
Imputer(add_indicator_features=True, axis=0, copy=True, missing_values=-1,
    strategy='mean', verbose=0)
>>> imputer.transform(X1)
array([[ 10.,   2.,   3.,   1.,   0.,   0.],
       [  4.,   6.,  12.,   0.,   0.,   1.],
       [  8.,  10.,  11.,   0.,   0.,   0.],
       [ 12.,   9.,  15.,   0.,   1.,   0.],
       [ 16.,  18.,  19.,   0.,   0.,   0.]])
>>> imputer.transform(X2)
array([[ 10.,   1.,   3.,   1.,   0.,   0.],
       [  4.,   0.,  12.,   0.,   0.,   1.],
       [  8.,   1.,   0.,   0.,   0.,   0.],
       [  0.,   0.,  15.,   0.,   0.,   0.],
       [ 16.,   1.,  19.,   0.,   0.,   0.]])

>>> imputer = Imputer(missing_values=-1, strategy='mean', axis=0,add_indicator_features=True)
>>> imputer.fit(X2)
Imputer(add_indicator_features=True, axis=0, copy=True, missing_values=-1,
    strategy='mean', verbose=0)
>>> imputer.transform(X1)
array([[  7.  ,   2.  ,   3.  ,   1.  ,  0.  ],
       [  4.  ,   6.  ,   9.25,   0.  ,   1.  ],
       [  8.  ,  10.  ,  11.  ,   0.  ,    0.  ],
       [ 12.  ,   0.6 ,  15.  ,   0.  ,   0.  ],
       [ 16.  ,  18.  ,  19.  ,   0.  ,   0.  ]])
>>> imputer.transform(X2)
array([[  7.  ,   1.  ,   3.  ,   1.  ,   0.  ],
       [  4.  ,   0.  ,   9.25,   0.  ,   1.  ],
       [  8.  ,   1.  ,   0.  ,   0.  ,   0.  ],
       [  0.  ,   0.  ,  15.  ,   0.  ,   0.  ],
       [ 16.  ,   1.  ,  19.  ,   0.  ,  0.  ]])
```

Is this the expected imputation since it depends on the missing values at fit time ?

jnothman commented:
Yes, that seems to be right.

On 15 May 2016 at 23:52, Maniteja Nandana <notifications@github.com> wrote:

> Sorry won't bother you more. I will try to show an example of what I
> understand here.
>
> >>> X1 = np.array([
> ...     [-1,  -1,   2,   3],
> ...     [4,  -1,   6,  -1],
> ...     [8,  -1,  10,  11],
> ...     [12,  -1,  -1,  15],
> ...     [16,  -1,  18,  19]
> ... ])
> >>>
> >>> X2 = np.array([
> ...     [-1,  -1,   1,   3],
> ...     [4,  -1,   0,  -1],
> ...     [8,  -1,   1,  0],
> ...     [0,  -1,   0,  15],
> ...     [16,  -1,   1,  19]
> ... ])
>
> >>> imputer = Imputer(missing_values=-1, strategy='mean', axis=0,add_indicator_features=True)
> >>> imputer.fit(X1)
> Imputer(add_indicator_features=True, axis=0, copy=True, missing_values=-1,
>     strategy='mean', verbose=0)
> >>> imputer.transform(X1)
> array([[ 10.,   2.,   3.,   1.,   0.,   0.],
>        [  4.,   6.,  12.,   0.,   0.,   1.],
>        [  8.,  10.,  11.,   0.,   0.,   0.],
>        [ 12.,   9.,  15.,   0.,   1.,   0.],
>        [ 16.,  18.,  19.,   0.,   0.,   0.]])
> >>> imputer.transform(X2)
> array([[ 10.,   1.,   3.,   1.,   0.,   0.],
>        [  4.,   0.,  12.,   0.,   0.,   1.],
>        [  8.,   1.,   0.,   0.,   0.,   0.],
>        [  0.,   0.,  15.,   0.,   0.,   0.],
>        [ 16.,   1.,  19.,   0.,   0.,   0.]])
>
> >>> imputer = Imputer(missing_values=-1, strategy='mean', axis=0,add_indicator_features=True)
> >>> imputer.fit(X2)
> Imputer(add_indicator_features=True, axis=0, copy=True, missing_values=-1,
>     strategy='mean', verbose=0)
> >>> imputer.transform(X1)
> array([[  7.  ,   2.  ,   3.  ,   1.  ,  0.  ],
>        [  4.  ,   6.  ,   9.25,   0.  ,   1.  ],
>        [  8.  ,  10.  ,  11.  ,   0.  ,    0.  ],
>        [ 12.  ,   0.6 ,  15.  ,   0.  ,   0.  ],
>        [ 16.  ,  18.  ,  19.  ,   0.  ,   0.  ]])
> >>> imputer.transform(X2)
> array([[  7.  ,   1.  ,   3.  ,   1.  ,   0.  ],
>        [  4.  ,   0.  ,   9.25,   0.  ,   1.  ],
>        [  8.  ,   1.  ,   0.  ,   0.  ,   0.  ],
>        [  0.  ,   0.  ,  15.  ,   0.  ,   0.  ],
>        [ 16.  ,   1.  ,  19.  ,   0.  ,  0.  ]])
>
> Is this the expected imputation since it depends on the missing values at
> fit time ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6782#issuecomment-219287028>
>

maniteja123 commented:
But if you look at transformed output of X2 then the feature with no missing values(feature 3) also has a indicator feature since it has missing value in X1, the input for ``fit``. 

And similarly for second case, the transformed output for X1 has no indicator feature for feature 3 though it has missing value since X2(input for ``fit``) didn't have missing value [This is the case you were suggesting earlier). 

Is that okay since we wanted only features with some missing values (but not everything)?

jnothman commented:
Yes, I think that's alright.

maniteja123 commented:
Thanks for clarifying. Will do the changes and push them in a short while once I get back home.

maniteja123 commented:
Hi sorry for pinging but I have done the changes to the best of my understanding. Please have a look if it is the expected behavior. Thanks.

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -114,7 +114,8 @@ class Imputer(BaseEstimator, TransformerMixin):
 
     imputed_features_ : array of shape (n_features_with_missing, )
-        The input features which have been imputed during transform.
+        This attribute is available only when ``axis=0``.
```
Can you put this after the following sentence, please?

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -189,6 +193,14 @@ def _sparse_fit(self, X, strategy, missing_values, axis):
+        if axis == 0:
+            mask = _get_mask(X.data, self.missing_values)    
+            mask_matrix = X.__class__((mask, X.indices.copy(),
+                                      X.indptr.copy()), shape=X.shape,
```
This should be indented by another space, to line up with `mask` above.

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -177,6 +178,9 @@ def fit(self, X, y=None):
                                                    self.axis)
+            invalid_mask = np.isnan(self.statistics_)
+            self.all_missing_idx_ = np.where(invalid_mask)[0]
+            self.imputed_features_ = np.setdiff1d(self.features_with_missing_values_, self.all_missing_idx_)
```
PEP8

maniteja123 commented:
Sorry the pep8 was old in my system which I failed to observe. I have updated it and fixed almost all except some where there are more than 79 characters. Right now , the shape of output should be independent of input shape of transform and dependent on fit only. Please have a look at your convenience and let me know if there is anything else to done. Thanks.

jnothman commented:
> except some where there are more than 79 characters

we do generally adhere to that requirement

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -311,37 +326,48 @@ def _dense_fit(self, X, strategy, missing_values, axis):
+                                          dtype=X.dtype)
+            inv_mask_matrix.eliminate_zeros()  # removes explicit False entries
+            all_missing_ = inv_mask_matrix.sum(axis=0).A.nonzero()[1]
+            self.mask_matrix_ = self.mask_matrix_[:, self.imputed_features_]
```
we should not be setting any attributes at transform time

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -311,37 +326,48 @@ def _dense_fit(self, X, strategy, missing_values, axis):
         values = np.repeat(valid_stats, n_missing)
+        if self.axis == 0:
+            all_missing_ = np.where(np.all(mask, axis=0))[0]
+            self.mask_ = self.mask_[:, self.imputed_features_]
```
we should not be setting any attributes at transform time

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -413,6 +431,21 @@ def transform(self, X):
+            mask = _get_mask(X.data, self.missing_values)
+            indexes = np.repeat(np.arange(len(X.indptr) - 1, dtype=np.int),
+                                np.diff(X.indptr))[mask]
+            self.mask_matrix_ = X.__class__((mask, X.indices.copy(),
```
we should not be setting any attributes at transform time

maniteja123 replied:
I remember that we should not set any attributes during transform time, then would it be okay to pass the ``mask_matrix_/mask_`` as an argument for ``sparse_transform_/dense_transform_`` from ``transform`` ?

jnothman replied:
Of course you can, but that's not to say it's the best or only way to do it.

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -311,37 +326,48 @@ def _dense_fit(self, X, strategy, missing_values, axis):
         indexes = np.repeat(np.arange(len(X.indptr) - 1, dtype=np.int),
                             np.diff(X.indptr))[mask]
+        if self.axis == 0:
+            not_mask = np.logical_not(mask)
```
Rather than duplicating this data, can't you just check for zero elements over the sum?

maniteja123 replied:
If we sum over the axis=0, zero elements will correspond to the features with no missing values. Would something like ``all_missing_ = np.where(np.all(mask_matrix_.A, axis=0))[0]`` be okay ? I am converting to dense array since there is no ``all`` implemented for sparse matrices.

jnothman replied:
`sum` with a boolean array is going to give you `all` pretty easily...

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -189,6 +194,14 @@ def _sparse_fit(self, X, strategy, missing_values, axis):
+            mask_matrix = X.__class__((mask, X.indices.copy(),
+                                       X.indptr.copy()), shape=X.shape,
+                                      dtype=X.dtype)
+            mask_matrix.eliminate_zeros()  # removes explicit False entries
```
If we're calling `sum` (and not doing anything else with `mask_matrix`), we don't get any benefit from this line. Remove it.

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -311,37 +326,48 @@ def _dense_fit(self, X, strategy, missing_values, axis):
+            inv_mask_matrix = X.__class__((not_mask, X.indices.copy(),
+                                           X.indptr.copy()), shape=X.shape,
+                                          dtype=X.dtype)
+            inv_mask_matrix.eliminate_zeros()  # removes explicit False entries
```
remove

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -413,6 +431,21 @@ def transform(self, X):
         valid_idx = np.where(valid_mask)[0]
         missing = np.arange(X.shape[not self.axis])[invalid_mask]
 
+        if sparse.issparse(X) and self.missing_values != 0:
```
why are you doing this and then doing `_{dense,sparse}_transform`?

maniteja123 replied:
I tried avoiding this, but the mask needed to be computed before the features with all missing values are discarded since the indicator matrix needs to select the columns based on ``imputed_features_`` which is with respect to the initial array before discarding features. Is there any way I can prevent this ?

jnothman commented on sklearn/preprocessing/tests/test_imputation.py:
```diff
@@ -386,23 +392,99 @@ def check_indicator(X, expected_imputed_features, axis):
+
+    X2 = np.array([
+        [-1,  -1,   1,   3],
+        [4,  1,   0,  -1],
```
Please add a comment to explain what warning this test is about...

jnothman commented on sklearn/preprocessing/tests/test_imputation.py:
```diff
@@ -386,23 +392,99 @@ def check_indicator(X, expected_imputed_features, axis):
+    assert_array_equal(Xt_with_in, np.hstack((Xt, imputed_features_mask)))
+    assert_array_equal(Xt_with_in,
+                       imputer_with_in.transform(sparse.csc_matrix(X2)).A)
+    assert_array_equal(Xt_with_in,
```
Is this assertion duplicated?

maniteja123 replied:
One is for csr_matrix while other is for csc_matrix. I suppose it is redundant but thought it will be better to have separate assertions for both.

jnothman replied:
Right. Yes, it's okay to have both.

jnothman commented on sklearn/preprocessing/tests/test_imputation.py:
```diff
@@ -386,23 +392,99 @@ def check_indicator(X, expected_imputed_features, axis):
+    assert_array_equal(Xt_with_in, np.hstack((Xt, imputed_features_mask)))
+    assert_array_equal(Xt_with_in,
+                       imputer_with_in.transform(sparse.csc_matrix(X2)).A)
+    assert_array_equal(Xt_with_in,
```
Ditto.

jnothman commented on sklearn/preprocessing/tests/test_imputation.py:
```diff
@@ -386,23 +392,99 @@ def check_indicator(X, expected_imputed_features, axis):
+    check_indicator(X, None, axis=1)
+
+
+def test_indicator_features_shape():
```
This seems a very poor name

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -413,6 +431,21 @@ def transform(self, X):
+                X = X.toarray()
+            self.mask_ = _get_mask(X, self.missing_values)
+
         if self.axis == 0 and invalid_mask.any():
```
A simple way to fix the code structure is to put this in a separate helper and call it from each of `_{sparse,dense}_transform`

maniteja123 commented on sklearn/preprocessing/imputation.py:
```diff
@@ -350,19 +380,28 @@ def _dense_transform(self, X, valid_stats, valid_idx):
 
         return X
 
+    def _preprocess(self, X, statistics):
```
Sorry I couldn't come up with a better name for this helper. Any suggestions would be helpful. Thanks.

jnothman replied:
How about `_handle_all_missing`

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -177,6 +178,10 @@ def fit(self, X, y=None):
                                                    self.strategy,
                                                    self.missing_values,
                                                    self.axis)
+            invalid_mask = np.isnan(self.statistics_)
```
Is this convention that `nan` marks all-missing columns something established by documentation and/or tests?

maniteja123 replied:
It isn't explicitly mentioned anywhere but if the column has all missing values, the corresponding element in ``statistics_`` is ``nan``.

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -177,6 +178,10 @@ def fit(self, X, y=None):
                                                    self.missing_values,
                                                    self.axis)
+            invalid_mask = np.isnan(self.statistics_)
+            self.all_missing_ = np.where(invalid_mask)[0]
```
If we're making this an underscore-suffixed attribute, it should be documented.

maniteja123 replied:
IMHO I don't know if this attribute would be of any great help but it needs it to be stored since the  transform function should know the features so that it can raise a ``RuntimeWarning`` in case that feature has some non-missing values in ``transform``.

jnothman replied:
then make it public; document it.

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -189,6 +194,13 @@ def _sparse_fit(self, X, strategy, missing_values, axis):
+            mask_matrix = X.__class__((mask, X.indices.copy(),
+                                       X.indptr.copy()), shape=X.shape,
+                                      dtype=X.dtype)
+            self.feat_with_missing_ = mask_matrix.sum(axis=0).A.nonzero()[1]
```
You really shouldn't be using attributes like this to transfer information back to a calling function. That's what return values are for. Storing this as an attribute is redundant with `imputed_features_`

maniteja123 replied:
Yeah sorry my bad. I will send it as a return value, but the reason I stopped from doing that was currently the return value ``statistic`` is sent at the end of evaluation of the code path below and also ``feat_with_missing_`` is only makes sense for ``axis=0``.

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -311,34 +325,50 @@ def _dense_fit(self, X, strategy, missing_values, axis):
-    def _sparse_transform(self, X, valid_stats, valid_idx):
+    def _sparse_transform(self, X, statistics):
         """transformer on sparse data."""
+        imputer_mask_ = _get_mask(X.data, self.missing_values)
```
why do these local variables end in underscore?

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -311,34 +325,50 @@ def _dense_fit(self, X, strategy, missing_values, axis):
-    def _dense_transform(self, X, valid_stats, valid_idx):
+    def _dense_transform(self, X, statistics):
         """transformer on dense data."""
+        imputer_mask_ = _get_mask(X, self.missing_values)
```
same here?

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -311,34 +325,50 @@ def _dense_fit(self, X, strategy, missing_values, axis):
+        X, valid_stats = self._preprocess(X, statistics)
+
+        if self.axis == 0:
+            all_missing_ = np.where(np.all(mask_matrix_.A, axis=0))[0]
```
No, the point of a sparse matrix is that densifying it would be a huge waste of memory. You need to find a way to not do that.

jnothman replied:
`sum`, for instance, should work fine to make a test of all being True.

maniteja123 replied:
yeah sorry I understand the memory wastage. How about something like ``all_missing = np.where(mask_matrix.sum(axis=0).A == X.shape[0])[0]`` ?

jnothman replied:
sounds good

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -377,7 +416,7 @@ def transform(self, X):
                 shape (n_samples, n_features_new) when
                 ``add_indicator_features`` is False,
-                shape (n_samples, n_features_new + len(imputed_features_)
+                shape (n_samples,n_features + n_features_new)
```
space after comma, please

jnothman commented:
This is very close to my +1. Apart from densifying that sparse matrix, it's mostly style issues.

maniteja123 commented:
Hi @jnothman, I have done the changes you suggested. But travis is failing on some tests, giving an error on ``np.setdiff1d``. It seems when one of the arrays is empty but am unable to figure out exactly why. Could you help please ?

jnothman commented:
The error is "arrays must have same number of dimensions". Is it possible that in `np.setdiff1d(self.all_missing_, all_missing)` one of these is 2d?

maniteja123 commented:
It is always ``all_missing_ = np.where(...)[0]`` so AFAICS I really don't think there is a possibility of that occurring and moreover tests are failing only on some builds. Would something like ``(0,)`` and ``(1,)`` have a dimensions mismatch because those are the sizes in some of the tests which are failing but I can't reproduce on my system ?

jnothman commented:
> It is always all_missing_ = np.where(...)[0] so AFAICS I really don't think there is a possibility of that occurring and moreover tests are failing only on some builds. 

Have you checked? sparse matrices can be ... special.

maniteja123 commented:
Yeah I did check the shape of both attributes in sparse and dense cases. They were 1D in case of all the tests.

jnothman commented:
I think it's associated with old scipy versions.

jnothman commented:
By which I mean it's a scipy bug that has since been fixed which causes `X.nonzero()` to return 2d matrices. They can be flattened (`.A.ravel()`), with a note that this fix supports old scipy (at least 0.11)

jnothman commented:
Could you please stop amending your commit? It's much easier to work with a commit history so that as a reviewer I can see what has changed. If we feel the need to squash a whole series of commits before merging, that's fine, but for reviewing, please make use of git rather than subverting it.

maniteja123 commented:
I am so sorry. I thought it would be easier to merge and was squashing them. Please correct me if I misunderstood your comment. Since ``X.nonzero()`` would return a matrix, ``.A.ravel()`` should  be done only after checking for version of scipy as 0.11 ? Thanks

jnothman commented:
I didn't mean my words to come across harshly. A few contributors have lately taken on the habit of squashing rather than adding a commit history, but it really can make review a lot harder.

No, you don't need to do a version check. We do have the option of adding `nonzero` to `sklearn.utils.fixes`, but I suspect that's overkill, especially since `.A.ravel()` from a matrix is such a light-weight operation. Just do it in all cases and add a comment that it's only needed to facilitate compatibility with old versions of scipy where `csr_matrix.nonzero()` returns matrices.

maniteja123 commented:
No no, it is perfectly fine. I understand how much of a responsibility this work is. Will try to take care from next time.

Regarding the fix, I tried adding ``ravel`` it to the output of ``nonzero`` but still the tests are failing. Could you please point me to the right place because I am unable to trace back where exactly the error is occurring ? Thanks.

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -189,6 +202,15 @@ def _sparse_fit(self, X, strategy, missing_values, axis):
+            mask_matrix = X.__class__((mask, X.indices.copy(),
+                                       X.indptr.copy()), shape=X.shape,
+                                      dtype=X.dtype)
+            feat_with_missing = mask_matrix.sum(axis=0).A.nonzero()[1]
```
It shouldn't be `.A.nonzero()`. We don't want to transform the sparse matrix to dense.

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -189,6 +202,15 @@ def _sparse_fit(self, X, strategy, missing_values, axis):
+                                      dtype=X.dtype)
+            feat_with_missing = mask_matrix.sum(axis=0).A.nonzero()[1]
+            # ravel since nonzero returns 2d matrices for sparse in scipy 0.11
+            feat_with_missing = feat_with_missing.ravel()
```
It should rather be `feat_with_missing.A.ravel()` to convert `feat_with_missing` from matrix (not sparse) to array, then flatten.

jnothman replied:
But if this didn't work, I'm not sure it's the entire problem.

maniteja123 replied:
Thanks, I got it now. But in the current version this will give an error since ``nonzero`` isn't returning ``matrix``.

jnothman replied:
Ah. Right. Don't use `.A` then. Sorry, not thinking at my best. `np.asarray` should convert matrix to array just as well.

maniteja123 committed:
Sparse fix

maniteja123 commented on sklearn/preprocessing/imputation.py:
```diff
@@ -309,36 +338,57 @@ def _dense_fit(self, X, strategy, missing_values, axis):
+            all_missing = np.where(mask_matrix.sum(axis=0) == X.shape[0])[0]
+            mask_matrix = mask_matrix[:, self.imputed_features_]
+            if not np.array_equal(all_missing, self.all_missing_):
+                features = np.setdiff1d(self.all_missing_, all_missing)
```
Sorry, I don't think I was looking at the right place. The tests are failing here. But ``nonzero`` isn't being used to get either of these.

jnothman replied:
So: use standard debugging techniques. It says they're different numbers of dimensions, so one of these is 2d (perhaps a `numpy.matrix`?). Where is each of them set?

maniteja123 replied:
The attribute ``self.all_missing_`` is set at [L190](https://github.com/scikit-learn/scikit-learn/pull/6782/files#diff-79ce5d2b4108efb44aa9de687d3af6e2R190) while ``all_missing_`` is set at [L357](https://github.com/scikit-learn/scikit-learn/pull/6782/files#diff-79ce5d2b4108efb44aa9de687d3af6e2R357)

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -112,9 +112,14 @@ class Imputer(BaseEstimator, TransformerMixin):
+    all_missing_ : array of shape (n_features - n_features_new, )
+        The features with all missing values for axis = 0
 
     imputed_features_ : array of shape (n_features_with_missing, )
```
I'm not sure this is a good name for it. Maybe `indicated_features_` or `incomplete_features_`. one problem is that nothing is imputed during fit.

maniteja123 replied:
I agree. From my knowledge, `incomplete_features_`` conveys the message clearly but just my opinion.

jnothman replied:
I think we'll make it `indicator_features_`. It's more explicit about what it's for.

maniteja123 committed:
try to fix sparse bug

maniteja123 committed:
convert matrix to ndarray

maniteja123 committed:
Fix tests

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -355,6 +355,8 @@ def _sparse_transform(self, X, statistics):
 
         if self.axis == 0:
             all_missing = np.where(mask_matrix.sum(axis=0) == X.shape[0])[0]
+            if isinstance(all_missing, np.matrix):
```
Please comment that this is for old scipy

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -112,9 +112,14 @@ class Imputer(BaseEstimator, TransformerMixin):
     ----------
     statistics_ : array of shape (n_features,)
         The imputation fill value for each feature if axis == 0.
+        The values corresponding to features ``all_missing_` is ``nan``.
```
This should be reworded, something like: "For a feature missing in all training samples, the value is nan".

From the code below, you assume that wherever the statistic is `nan`, it means all are missing. I'm not sure that's entirely true... If it is, what do we gain from having a separate `all_missing_`?

maniteja123 replied:
Sorry for the delay. Thanks will reword it. You are right about the statistic being ``nan`` when all are missing. Even if there is one value for the feature, the corresponding statistic becomes the value itself. I am using ``all_missing_`` to keep a track of features with all missing values during fit, so that it is possible to compare with the input to transform and raise a RunTime Warning accordingly. Please do let me know if there is a better way to achieve that.

jnothman commented on sklearn/preprocessing/imputation.py:
```diff
@@ -377,7 +438,7 @@ def transform(self, X):
                 shape (n_samples, n_features_new) when
                 ``add_indicator_features`` is False,
-                shape (n_samples, n_features_new + len(imputed_features_)
+                shape (n_samples, n_features + n_features_new)
```
I don't think this is clear, or accurate. What is `n_features_new`?

maniteja123 replied:
Sorry for the poor choice of words. In case a whole feature is discarded due to all missing values, ``n_features_new`` will be the number of features in the imputer output without indicator_features. In that context, it should be ``n_features_new + len(imputed_features_)``. ``n_features_new`` might not be equal to ``len(imputed_features_)`` in the case of ``axis=0``, since ``imputed_features_`` can discard the features

jnothman commented on sklearn/preprocessing/tests/test_imputation.py:
```diff
@@ -383,26 +389,105 @@ def check_indicator(X, expected_imputed_features, axis):
+    imputer = Imputer(missing_values=-1, strategy='mean', axis=0)
+    imputer.fit(X1)
+    imputer.transform(X2)
+    assert_warns(RuntimeWarning, imputer.transform, X2)
```
Please use `assert_warns_message` to make the assertion more explicit.

jnothman commented:
Thanks @maniteja123. Apart from not being sure that it's sufficiently documented, this is looking pretty good to what we specified. Even so (sorry @mfeurer), I continue to be undecided on whether it's the *correct* approach.

I'm now wondering, for instance: why, if we're only adding indicators for features with training-time missing values when axis=0, are we not doing the same for `axis=1`? After all, that restriction is really about features being useless if they're constant at training time. Maybe the `axis=1` case should use `imputed_features_` identically to `axis=0`.

Alternatively, we could just default to including *all* features in the indicator matrix, unless specified by providing a list of feature indices like `add_indicator_features=[0, 2, 3]`. In that case, we have no need for `imputed_features_`.

jnothman commented:
@maniteja123, I'd like to make some changes to this myself, in part because I keep changing the spec, in part because I'm a little anxious to make sure this makes it into 0.18 (or else IMO we'll need to revert `add_indicator_features`). It's now looking close, but there are still some bits and pieces that could be neatened up. Is it okay if I offer pull requests to your branch? Alternatively, I could make a new PR from my own branch, and take ownership of the process of fixing it up for merge (either way we'll need two others to review).

maniteja123 commented:
Sure Joel, please do take over the work in case you wish to. I understand that this needs to come to a conclusive state before the release. Sorry for being unable to complete it in time. Do let me know if I can do anything to assist.

raghavrv commented:
Any updates on this? I need this for my benchmarks...

raghavrv commented on sklearn/preprocessing/imputation.py:
```diff
@@ -309,36 +338,59 @@ def _dense_fit(self, X, strategy, missing_values, axis):
+                features = np.setdiff1d(self.all_missing_, all_missing)
+                warnings.warn("The features %s have all missing "
+                              "values in fit but have non-missing values"
+                              "in transform " % features, RuntimeWarning,
```
space before `in`

raghavrv commented on sklearn/preprocessing/imputation.py:
```diff
@@ -309,36 +338,59 @@ def _dense_fit(self, X, strategy, missing_values, axis):
+                features = np.setdiff1d(self.all_missing_, all_missing)
+                warnings.warn("The features %s have all missing"
+                              " values in fit but have non-missing values in"
+                              "transform" % features, RuntimeWarning,
```
space before `transform`

amueller commented:
@raghavrv wait it this PR still relevant? The missing feature support was completely removed from master by @jnothman and needs a rewrite.

raghavrv commented:
This wasn't the rewrite?

raghavrv commented:
My bad this simply tries to address the issue raised by @mfeurer... Should we close this then?

amueller commented:
If this is a patch against the old implementation, which I think it is, then yes. @maniteja123 can you confirm?

raghavrv commented:
Indeed it is.

maniteja123 commented:
Hi yeah it is based on old implementation. I suppose it can be closed. Sorry for keeping it open.

jnothman commented:
Well it can be opened and fixed up, or a new PR opened. it's fine for it to
be based on the old from a code perspective: GH'll take the appropriate
diff. But the discussion here is less relevant

On 14 October 2016 at 06:00, Maniteja Nandana <notifications@github.com>
wrote:

> Closed #6782 <https://github.com/scikit-learn/scikit-learn/pull/6782>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6782#event-822970807>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz61eG_h1JAyD6suuzZ-mG0pan4p-0ks5qzn_SgaJpZM4IephV>
> .
>

raghavrv commented:
@maniteja123 Are you up for it?

maniteja123 commented:
Hi @raghavrv, I would be happy to work on it.

raghavrv commented:
Thanks please ping me for a review when you have a PR :)
